libraries/botbuilder-dialogs/src/dialogSet.ts

```diff
@@ -23,33 +23,34 @@ import { DialogContext, DialogState } from './dialogContext';
  * for their name and phone number:
  *
  * ```JavaScript
- * const { DialogSet, TextPrompt } = require('botbuilder-dialogs');
+ * const { DialogSet, TextPrompt, WaterfallDialog } = require('botbuilder-dialogs');
  *
  * const dialogs = new DialogSet();
  *
- * dialogs.add('fillProfile', [
- *     async function (dc, options) {
+ * dialogs.add(new WaterfallDialog('fillProfile', [
+ *     async (dc, step) => {
  *         dc.activeDialog.state.profile = {};
- *         await dc.prompt('textPrompt', `What's your name?`);
+ *         return await dc.prompt('textPrompt', `What's your name?`);
  *     },
- *     async function (dc, name) {
- *         dc.activeDialog.state.profile.name = name;
- *         await dc.prompt('textPrompt', `What's your phone number?`);
+ *     async (dc, step) => {
+ *         dc.activeDialog.state.profile.name = step.results;
+ *         return await dc.prompt('textPrompt', `What's your phone number?`);
  *     },
- *     async function (dc, phone) {
- *         dc.activeDialog.state.profile.phone = phone;
+ *     async (dc, step) => {
+ *         dc.activeDialog.state.profile.phone = step.results;
  *
  *         // Save completed profile to user state
- *         const user = userState.get(context);
+ *         const user = await userState.get(context);
  *         user.profile = dc.activeDialog.state.profile;
+ *         await userState.set(context, user);
```

Comment:
For best practice, we should use state property accessor and avoid using the `set()` method because this method will delete the entire state before setting the new value to the state...it could have unexpected behavior if you have other data saved to the conversation or user state...

Reply:
The variable name could be clearer, but I believe this _is_ showing the use of a state property accessor.  It just happens to be named userState.  Am I wrong?

Reply:
This state is not using the accessor...it's using the botstate's get method which returns a cache state...This is generally a bad practice if you have multiple pieces of information stored in the userstate...like a "user info" bag and a "reservation" bag...when you call set in this statement, the user state is cleared then it saves whatever is in "user" to the state...not what a dev would expect (unless that is their intent, to clear all state and start with just the "user" data.

Using the state accessor:
```javascript
const conversationState = new ConversationState(storage);
const convoStateAccessor = conversationState.createProperty('convoState');
```

you can avoid this problem of deleting other data you stored because the accessor only grags you the property you want to work on...

`const convoState = await convoStateAccessor.get(context, {});`

you can then set values by individual properties or call the "set" method to replace whatever is in that property with your new one.

Hope that helps...

Reply:
I have updated this example to include clear references to setting up memory storage, user state, and a property for the values we are using.